### PR TITLE
started providing the GPU sync switch to external view embedders

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

External view embedders may be doing work on the GPU and they have to make sure they don't do so while the app is in the background.

I considered first making this a parameter to the ios_external_view_embedder but decided against it because:
1) If this external embedder needs to consider the gpu switch than another implementation of the ExternalViewEmbedder would probably have to consider it.
1) Creation of the iOS external view embedder happens before the shell is created so we'd have to use a setter instead of a constructor argument and there is the chance to introduce bugs.

## Related Issues

fixes https://github.com/flutter/flutter/issues/69642

## Tests

I added the following tests:

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Review by cyanglaz — COMMENTED** [review]
Code looks good. I have no clue what does `gpu_disable_sync_switch` do and how it works. Let's add another reviewer who knows this better?

**Comment by chinmaygarde** [comment]
cc @jason-simmons

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 1]
```diff
@@ -466,9 +466,22 @@
...
-                                                 std::shared_ptr<IOSContext> ios_context,
-                                                 std::unique_ptr<SurfaceFrame> frame) {
+bool FlutterPlatformViewsController::SubmitFrame(
+    GrDirectContext* gr_context,
+    std::shared_ptr<IOSContext> ios_context,
+    std::unique_ptr<SurfaceFrame> frame,
+    const std::shared_ptr<fml::SyncSwitch>& gpu_disable_sync_switch) {
+  bool result = false;
+  gpu_disable_sync_switch->Execute(
+      fml::SyncSwitch::Handlers().SetIfTrue([&] { result = false; }).SetIfFalse([&] {
```
Comment (jason-simmons) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:476]: Is the `SetIfTrue` handler necessary?  `result` is already set to `false` by default

**Review by jason-simmons — APPROVED** [review]

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:472]: Nope, it shouldn't be necessary.  Once I get things sorted out I'll try removing it.

Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:472]: Oh, shoot I forgot to do this before it landed.  I was in a rush.  It's probably not a big deal, if you want it gone I can do it in a separate PR.  Sorry.

**Comment by blasten** [comment]
@gaaclarke , @cyanglaz -  I came across this change. Do you know why not clearing the current pending raster tasks in `Shell.OnPlatformViewDestroyed`? 

`Shell.OnPlatformViewDestroyed` is  called from the same lifecycle event `applicationWillResignActive`
https://github.com/flutter/engine/blob/a598f23763cec207d7ccfa186f5f0280b30147df/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#L707

**Comment by gaaclarke** [comment]
@blasten If we synchronously flushed all outstanding GPU tasks on the raster thread for the external view embedder when we got the ResignActive notification and guaranteed that it will never attempt another GPU tasks until the app has become active again that would be a valid way to solve this issue, too.  Is that what you are asking?

**Comment by blasten** [comment]
@gaaclarke yep. @cyanglaz mentioned that the crash would happen here https://github.com/flutter/engine/blob/master/shell/common/rasterizer.cc#L438 because we teared down the surface from the link I mentioned above.

It sounds like we need a primitive to ensure that all tasks are flushed and that the queue is disabled.

**Comment by gaaclarke** [comment]
@blasten Check out the issue this PR addresses, the stacktrace of the crash is there.  It's from the external view embedder executing GPU operations.  This PR causes the platform thread to wait for GPU operations to execute before exiting, then turns them off atomically.  Any other attempts at creating frames in the background will be dropped.  It's the same mechanism we use elsewhere like the decoding of images on background threads.

**Comment by blasten** [comment]
right. I guess my question is more about defining some primitives that work well across platforms. Do you know if we need to this for Android too?

**Comment by gaaclarke** [comment]
I believe iOS is the only platform we currently support that has the requirement that GPU access may be blocked.  Android at least doesn't crash, for energy reason we might want to disable GPU operations in the background if the operating system isn't already handling that gracefully.

**Pull request merged**